src/support/mtx_rw.c

```diff
@@ -198,37 +163,75 @@ __wt_readlock_spin(WT_SESSION_IMPL *session, WT_RWLOCK *l)
 void
 __wt_readlock(WT_SESSION_IMPL *session, WT_RWLOCK *l)
 {
-	uint16_t ticket;
+	WT_RWLOCK new, old;
 	int pause_cnt;
+	uint8_t ticket;
 
 	WT_STAT_CONN_INCR(session, rwlock_read);
 
 	WT_DIAGNOSTIC_YIELD;
 
-	/*
-	 * Possibly wrap: if we have more than 64K lockers waiting, the ticket
-	 * value will wrap and two lockers will simultaneously be granted the
-	 * lock.
-	 */
-	ticket = __wt_atomic_fetch_add16(&l->s.next, 1);
-	for (pause_cnt = 0; ticket != l->s.readers;) {
+	pause_cnt = 0;
+	do {
+restart:	/*
+		 * Fast path: if there is no active writer, join the current
+		 * group.
+		 */
+		for (old.u.v = l->u.v;
+		    old.u.s.current == old.u.s.next;
+		    old.u.v = l->u.v) {
+			new.u.v = old.u.v;
+			new.u.s.readers_active++;
+			if (__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v))
+				return;
+			WT_PAUSE();
+		}
+
+		/*
+		 * There is an active writer: join the next group.
+		 * Check for wrapping: if the maximum number of readers are
+		 * already queued, wait until we can get a valid ticket.
+		 * If we are the first reader to queue, set the next read
+		 * group.
+		 * Note: don't re-read from the lock or we could race with a
+		 * writer unlocking.
+		 */
+		if (old.u.s.readers_queued == UINT16_MAX ||
+		    (old.u.s.next != old.u.s.current &&
+		    old.u.s.readers_queued > old.u.s.next - old.u.s.current)) {
+			__wt_cond_wait(
+			    session, l->cond_readers, WT_THOUSAND, NULL);
+			goto restart;
+		}
+		new.u.v = old.u.v;
+		if (new.u.s.readers_queued++ == 0)
+			new.u.s.reader = new.u.s.next;
+		ticket = new.u.s.reader;
+
 		/*
-		 * We failed to get the lock; pause before retrying and if we've
-		 * paused enough, yield so we don't burn CPU to no purpose. This
-		 * situation happens if there are more threads than cores in the
-		 * system and we're thrashing on shared resources.
+		 * Check for wrapping: if we have more than 64K lockers
+		 * waiting, the ticket value will wrap and two lockers will
+		 * simultaneously be granted the lock.
 		 */
-		if (++pause_cnt < WT_THOUSAND)
+		WT_ASSERT(session, new.u.s.readers_queued != 0);
+	} while (!__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v));
```

Comment:
The `Check for wrapping: if we have more than 64K lockers` comment no longer applies to this code.

I think we can remove the assert on `readers_queued != 0`, we know it's safe, we incremented it immediately before testing it.